File: VizOs2/VizOs/frontend/js/fcfs.js
// First Come First Serve (FCFS) CPU Scheduling Algorithm
class FCFS {
    static simulate(processes) {
        // Sort processes by arrival time
        const sortedProcesses = [...processes].sort((a, b) => a.arrival - b.arrival);
        
        let currentTime = 0;
        const ganttChart = [];
        const processResults = [];
        
        // Calculate completion time, turnaround time, waiting time, and response time
        sortedProcesses.forEach(process => {
            const startTime = Math.max(currentTime, process.arrival);
            const completionTime = startTime + process.burst;
            const turnaroundTime = completionTime - process.arrival;
            const waitingTime = turnaroundTime - process.burst;
            const responseTime = startTime - process.arrival;
            
            // Add to Gantt chart
            ganttChart.push({
                name: process.id,
                startTime: startTime,
                duration: process.burst
            });
            
            // Store process results
            processResults.push({
                id: process.id,
                arrivalTime: process.arrival,
                burstTime: process.burst,
                startTime: startTime,
                completionTime: completionTime,
                turnaroundTime: turnaroundTime,
                waitingTime: waitingTime,
                responseTime: responseTime
            });
            
            currentTime = completionTime;
        });
        
        // Calculate metrics
        const totalWaitingTime = processResults.reduce((sum, p) => sum + p.waitingTime, 0);
        const totalTurnaroundTime = processResults.reduce((sum, p) => sum + p.turnaroundTime, 0);
        const totalResponseTime = processResults.reduce((sum, p) => sum + p.responseTime, 0);
        const totalBurstTime = processResults.reduce((sum, p) => sum + p.burstTime, 0);
        
        const avgWaitingTime = totalWaitingTime / processResults.length;
        const avgTurnaroundTime = totalTurnaroundTime / processResults.length;
        const avgResponseTime = totalResponseTime / processResults.length;
        const cpuUtilization = totalBurstTime / currentTime;
        
        return {
            algorithm: 'FCFS',
            ganttChart: {
                processes: ganttChart,
                totalTime: currentTime
            },
            processResults: processResults,
            metrics: {
                avgWaitingTime: avgWaitingTime,
                avgTurnaroundTime: avgTurnaroundTime,
                avgResponseTime: avgResponseTime,
                cpuUtilization: cpuUtilization
            }
        };
    }
}

File: VizOs2/VizOs/algorithms/FCFS.py
"""
FCFS (First Come First Serve) Algorithm
Non-preemptive CPU scheduling algorithm
"""

def execute(processes):
    """
    Execute FCFS scheduling algorithm
    
    Args:
        processes: List of process dictionaries with:
            - id: Process identifier
            - arrival_time: Arrival time
            - burst_time: CPU burst time
            
    Returns:
        Dictionary containing:
            - gantt_chart: List of execution segments
            - metrics: Performance metrics
            - process_details: Detailed process information
    """
    # Sort processes by arrival time
    sorted_processes = sorted(processes, key=lambda x: x['arrival_time'])
    
    # Initialize variables
    gantt_chart = []
    process_details = []
    current_time = 0
    
    # Execute processes in order
    for proc in sorted_processes:
        arrival = proc['arrival_time']
        burst = proc['burst_time']
        
        # Wait if process hasn't arrived
        if current_time < arrival:
            current_time = arrival
        
        # Calculate times
        waiting_time = current_time - arrival
        turnaround_time = waiting_time + burst
        
        # Add to Gantt chart
        gantt_chart.append({
            'process': proc['id'],
            'start': current_time,
            'end': current_time + burst
        })
        
        # Update process details
        process_details.append({
            'id': proc['id'],
            'arrival_time': arrival,
            'burst_time': burst,
            'waiting_time': waiting_time,
            'turnaround_time': turnaround_time,
            'completion_time': current_time + burst
        })
        
        # Update current time
        current_time += burst
    
    # Calculate metrics
    total_waiting_time = sum(p['waiting_time'] for p in process_details)
    total_turnaround_time = sum(p['turnaround_time'] for p in process_details)
    n = len(process_details)
    
    metrics = {
        'average_waiting_time': round(total_waiting_time / n, 2),
        'average_turnaround_time': round(total_turnaround_time / n, 2),
        'cpu_utilization': round((sum(p['burst_time'] for p in process_details) / current_time) * 100, 2) if current_time > 0 else 0,
        'throughput': round(n / current_time, 2) if current_time > 0 else 0
    }
    
    return {
        'gantt_chart': gantt_chart,
        'metrics': metrics,
        'process_details': process_details
    }

